src/Build/Graph/ProjectInterpretation.cs

```diff
@@ -0,0 +1,387 @@
+                    break;
+                case ProjectType.NonMultitargeting:
+                    globalPropertiesModifiers = ModifierForNonMultitargetingNodes;
+                    references = requesterInstance.GetItems(ItemTypeNames.ProjectReference);
+                    break;
+                default:
+                    throw new ArgumentOutOfRangeException();
+            }
+
+            foreach (var projectReference in references)
+            {
+                if (!String.IsNullOrEmpty(projectReference.GetMetadataValue(ToolsVersionMetadataName)))
+                {
+                    throw new InvalidOperationException(
+                        String.Format(
+                            CultureInfo.InvariantCulture,
+                            ResourceUtilities.GetResourceString(
+                                "ProjectGraphDoesNotSupportProjectReferenceWithToolset"),
+                            projectReference.EvaluatedInclude,
+                            requesterInstance.FullPath));
+                }
+
+                var projectReferenceFullPath = projectReference.GetMetadataValue(FullPathMetadataName);
+
+                var referenceGlobalProperties = GetGlobalPropertiesForItem(projectReference, requesterInstance.GlobalPropertiesDictionary, globalPropertiesModifiers);
+
+                var referenceConfig = new ConfigurationMetadata(projectReferenceFullPath, referenceGlobalProperties);
+
+                yield return referenceConfig;
+            }
+        }
+
+        private static string GetInnerBuildPropertyValue(ProjectInstance project)
+        {
+            return project.GetPropertyValue(GetInnerBuildPropertyName(project));
+        }
+
+        private static string GetInnerBuildPropertyName(ProjectInstance project)
+        {
+            return project.GetPropertyValue(PropertyNames.InnerBuildProperty);
+        }
+
+        private static string GetInnerBuildPropertyValues(ProjectInstance project)
+        {
+            return project.GetPropertyValue(project.GetPropertyValue(PropertyNames.InnerBuildPropertyValues));
+        }
+
+        internal static ProjectType GetProjectType(ProjectInstance project)
+        {
+            var isOuterBuild = String.IsNullOrWhiteSpace(GetInnerBuildPropertyValue(project)) && !String.IsNullOrWhiteSpace(GetInnerBuildPropertyValues(project));
+            var isInnerBuild = !String.IsNullOrWhiteSpace(GetInnerBuildPropertyValue(project));
+
+            ErrorUtilities.VerifyThrow(!(isOuterBuild && isInnerBuild), $"A project cannot be an outer and inner build at the same time: ${project.FullPath}");
+
+            return isOuterBuild
+                ? ProjectType.OuterBuild
+                : isInnerBuild
+                    ? ProjectType.InnerBuild
+                    : ProjectType.NonMultitargeting;
+        }
+
+        public static void PostProcess(ConcurrentDictionary<ConfigurationMetadata, ProjectGraphNode> allNodes)
```

Comment:
```suggestion
        public static void PromoteInnerBuildReferencesToReferencingProjects(ConcurrentDictionary<ConfigurationMetadata, ProjectGraphNode> allNodes)
```

? If this is the way to keep it.

Reply:
Not sure, depends how the abstractions turn out. Right now I tried to keep ProjectGraph pretty agnostic of xml contents. It does not even know about multitargeting. ProjectInterpretation is the one that knows about evaluation particularities, what multitargeting is, etc. The vague intent is to potentially decouple ProjectInterpretation into plugins provided by SDKs.